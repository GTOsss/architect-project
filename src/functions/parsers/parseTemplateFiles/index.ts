export * from './parseTemplateFiles';
export * from './parseTemplateFiles.types';
export * from './parseTemplateFragments.types';
