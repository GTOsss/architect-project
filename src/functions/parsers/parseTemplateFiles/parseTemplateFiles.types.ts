import { RequiredTemplateScript } from '../../getters';
import { ParsedFragmentData } from './parseTemplateFragments.types';

export type ParsedTemplateFragment = {
  /** Original interpolation fragment without brackets */
  str: string;
  /** Original interpolation fragment with brackets */
  originStr: string;
  index: number;
  data: ParsedFragmentData;
};

/**
 * @example
 * {
 *     file: 'C:/project/architect/templates/component[name][name].tsx.hbs',
 *     parsed: [
 *       {
 *         str: 'name',
 *         index: 67,
 *         data: ParsedFragmentData,
 *       },
 *       {
 *         str: 'toCamelCase(name)',
 *         index: 97,
 *         data: ParsedFragmentData,
 *       },
 *     ],
 *     content: `
 *        import React, { memo } from 'react';
 *        import styles from './{{name}}.module.css';
 *
 *        type {{toCamelCase(name)}}Props = {} ...`,
 *   },
 * */
export type ParsedTemplateFile = {
  file: string;
  parsed: ParsedTemplateFragment[];
  content: string;
};

export type ParsedTemplateInfo = {
  templateScript: RequiredTemplateScript;
  parsedFiles: ParsedTemplateFile[];
};

export type ParsedTemplateMap = {
  [templateName: string]: ParsedTemplateInfo;
};

export enum ParserContextEnum {
  'filePath' = 'filePath',
  'fileContent' = 'fileContent',
}
