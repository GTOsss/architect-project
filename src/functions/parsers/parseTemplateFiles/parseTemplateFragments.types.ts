export enum ParsedFragmentDataTypeEnum {
  'variable' = 'variable',
  'callFunction' = 'callFunction',
}

export type ParsedFragmentVariable = {
  type: ParsedFragmentDataTypeEnum.variable;
  variableName: string;
};

export const isVariableData = (data: ParsedFragmentData): data is ParsedFragmentVariable =>
  data.type === ParsedFragmentDataTypeEnum.variable;

export type ParsedFragmentCallFunction = {
  type: ParsedFragmentDataTypeEnum.callFunction;
  functionName: string;
  functionArgs: string[];
};

export const isCallFunctionData = (data: ParsedFragmentData): data is ParsedFragmentCallFunction =>
  data.type === ParsedFragmentDataTypeEnum.callFunction;

export type ParsedFragmentData = ParsedFragmentVariable | ParsedFragmentCallFunction;
