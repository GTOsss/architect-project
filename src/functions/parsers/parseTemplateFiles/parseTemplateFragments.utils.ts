import { ParsedFragmentData, ParsedFragmentDataTypeEnum } from './parseTemplateFragments.types';

export const getFunctionName = (str: string): string => {
  const reGetFunction = /.+(?=\()/gm;
  return str.match(reGetFunction)[0];
};

export const getFunctionArgument = (str: string): string => {
  const reGetFunctionArgument = /(?<=\().+(?=\))/gm;
  return str.match(reGetFunctionArgument)?.[0];
};

/**
 * Return true if str look like call function.
 *
 * isCallFunction(a, b) // false // Todo Should be expanded to correct work with multi arguments
 * isCallFunction(1name) // true // Todo Need handle this case
 *
 * @example
 * isCallFunction('fn()') // true
 * isCallFunction('fn(name)') // true
 *
 * isCallFunction('some (some text)') // false
 * isCallFunction('some text (with brackets)') // false
 *
 * */
export const isCallFunctionStr = (str: string): boolean => {
  const reIsFunction = /[a-zA-Z0-9]+\s*\([a-zA-Z0-9]*\)/gm;
  return reIsFunction.test(str);
};

/**
 * Detect type of interpolation fragment and return parsed data of it.
 * */
export const getParsedFragmentData = (str: string): ParsedFragmentData => {
  if (isCallFunctionStr(str)) {
    const functionName = getFunctionName(str);
    const functionArg = getFunctionArgument(str);

    return { type: ParsedFragmentDataTypeEnum.callFunction, functionName, functionArgs: [functionArg] };
  }

  return { type: ParsedFragmentDataTypeEnum.variable, variableName: str };
};
