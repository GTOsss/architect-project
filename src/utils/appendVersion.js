const fs = require('file-system');
const configPath = require('../configPath');
const { resolve } = require('path');
const exportPath = resolve(configPath.arcHistoryPath, 'versions.json');

const createVersionsJson = () => {
  const defContent = {
    versions: [],
    current: 0,
  };
  const prettyDefContent = JSON.stringify(defContent, null, '  ');

  fs.writeFileSync(exportPath, prettyDefContent);

  return require(configPath.versionsJsonPath);
};

const json = fs.existsSync(configPath.versionsJsonPath) ? require(configPath.versionsJsonPath) : createVersionsJson();

const appendVersion = (date) => {
  json.versions.push(date);
  json.current = json.versions.length;

  const content = JSON.stringify(json, null, '  ');

  fs.writeFileSync(exportPath, content);
};

module.exports = appendVersion;
