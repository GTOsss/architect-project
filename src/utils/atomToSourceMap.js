const sourceMapByAtoms = {
  users: [['component', { path: 'src/custompath', a: 'a' }], 'types', 'store'],
  friends: ['component', ['store', { path: 'src/custompath' }]],
};

const defaultParams = {
  component: { path: 'src/components', b: 'b' },
  types: { path: 'src/types' },
  store: { path: 'src/store' },
};

// output
// const map = {
//   'src/components': {
//     users: component,
//   },
//   'src/types': {
//     users: types,
//   },
//   'src/store': {
//     users: store,
//   },
// };

let map = {};

const atomToSourceMap = ({ sourceMapByAtoms, defaultParams: allDefaultParams }) => {
  Object.entries(sourceMapByAtoms).forEach(([componentName, templates]) => {
    templates.forEach((template) => {
      const templateIsString = typeof template === 'string';

      const params = templateIsString ? {} : template[1];

      const templateName = templateIsString ? template : template[0];

      const defaultParams = allDefaultParams[templateName];

      const mergedParams = { ...defaultParams, ...params };

      console.log(mergedParams);

      const countParams = Object.keys(mergedParams).length;
      const paramsIsOnlyTemplateName = countParams > 1;

      const valueInSourceMap = paramsIsOnlyTemplateName ? componentName : { mergedParams };

      map[mergedParams.path] = map[mergedParams.path] ? map[mergedParams.path] : {};
      map[mergedParams.path] = { ...map[mergedParams.path], [templateName]: valueInSourceMap };
    });
  });

  // console.log(map);
};

atomToSourceMap({ sourceMapByAtoms, defaultParams });

//console.log(map);

// const sourceMapByAtoms = {
//   users: ['component', 'types', 'store'],
//   friends: ['component', 'types', 'store'],
// };

// Object.entries(sourceMapByAtoms).forEach(([components, templates]) => {
//   Object.entries(defaultParams).forEach(([defTemplates, defParams]) => {
//     if (templates.includes(defTemplates)) {
//       templates.forEach((template) => {
//         if (defTemplates === template) {
//           map = {
//             ...map,
//             [defParams.path]: {
//               [components]: template,
//             },
//           };
//         }
//       });
//     }
//   });
// });

// const atomToSourceMap = ({ sourceMapByAtoms, defaultParams }) => {
//   Object.entries(sourceMapByAtoms).forEach(([key, templates]) => {
//     templates.forEach((template) => {
//       if (typeof template === 'string') {
//         let params = defaultParams[template];
//         map = {
//           ...map,
//           [params.path]: {
//             [key]: template,
//           },
//         };
//       } else {
//         let params = template[1];
//         const value = {
//           [key]: template[0],
//         };
//         map = {
//           ...map,
//           [params.path]: value,
//         };
//       }
//       //  console.log(`current template: ${template}, params: `, params);
//     });
//   });
