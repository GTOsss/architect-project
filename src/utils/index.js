const getSources = (dirName) => {
  console.log('getSources dirName');
};

module.exports = {
  getSources,
};
