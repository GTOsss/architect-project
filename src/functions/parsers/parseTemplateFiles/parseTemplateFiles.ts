import fs from 'file-system';
import { TemplateObjWithPaths } from '../../getters/getObjectWithPaths.types';
import {
  ParsedTemplateFile,
  ParsedTemplateFragment,
  ParsedTemplateMap,
  ParserContextEnum,
} from './parseTemplateFiles.types';
import { getParsedFragmentData } from './parseTemplateFragments.utils';
import { ArcConfig } from '../../../types/config';

/**
 * Executes a regular expression on a given string and returns an array of matches.
 *
 * • Iterates over all matches of regexp.
 * • For each match found, it stores an object containing the matched substring and the zero-based index at which it was found.
 *
 * @param str - The string on which to execute the regular expression.
 * @param re - The regular expression to execute. Should have the global flag set
 *              (`/g`) for multiple matches, since `exec` advances the lastIndex property.
 * @param config - Arc config
 * @param context - Indicator of file content or path. Need for correct make "originStr"
 * @returns An array of objects, each representing a match. Each object contains:
 *          - `str`: the matched substring.
 *          - `index`: the starting position of the match within `str`.
 *
 * @example
 *
 * const str = "This is a {test} string with {multiple} braces.";
 * const re = /(?<=\{).+?(?=})/g; // lookbehind `(?<=\{)` and lookahead `(?=})` to find content enclosed in curly braces.
 * const matches = reExec(str, re); // result:
 *  [
 *    { data: ParsedFragmentData, str: 'test', originStr: '{{test}}',  index: 10 },
 *    { data: ParsedFragmentData, str: 'multiple', originStr: '{{multiple}}', index: 28 }
 *  ]
 *
 */
export const parseAllInterpolationMarks = (
  str: string,
  re: RegExp,
  config: ArcConfig,
  context: ParserContextEnum = ParserContextEnum.fileContent,
) => {
  const results: ParsedTemplateFragment[] = [];
  let result: RegExpExecArray;
  while ((result = re.exec(str))) {
    const currentResult = result[0];
    const data = getParsedFragmentData(currentResult);
    results.push({
      str: currentResult,
      originStr: getInterpolationOriginFragment(currentResult, config, context),
      index: result.index,
      data,
    });
  }
  return results;
};

export const getInterpolationOriginFragment = (str: string, config: ArcConfig, context: ParserContextEnum) => {
  if (context === ParserContextEnum.fileContent) {
    return [config.itrStart, str, config.itrEnd].join('');
  }

  return [config.itrFileNameStart, str, config.itrFileNameEnd].join('');
};

export const createInterpolationRegExp = (config: ArcConfig) =>
  new RegExp(`(?<=${config.itrStart}).+?(?=${config.itrEnd})`, 'gm');

/**
 * Iterates by template files, read and parse it to object with interpolation info.
 *
 * @example
 * // example of return data:
 * const result = [
 *   {
 *     file: 'C:/project/architect/templates/component[name][name].tsx.hbs',
 *     parsed: [
 *       {
 *         str: 'name',
 *         originStr: '{{name}}',
 *         index: 67,
 *         data: ParsedFragmentData,
 *       },
 *       {
 *         str: 'toCamelCase(name)',
 *         originStr: '{{toCamelCase(name)}}',
 *         index: 97,
 *         data: ParsedFragmentData,
 *       },
 *     ],
 *     content: `
 *        import React, { memo } from 'react';
 *        import styles from './{{name}}.module.css';
 *
 *        type {{toCamelCase(name)}}Props = {} ...`,
 *   },
 *   ...,
 * ];
 * */
export const parseTemplateFiles = (templates: TemplateObjWithPaths[], config: ArcConfig) => {
  const reInterpolation = createInterpolationRegExp(config);

  const parsedTemplateMap: ParsedTemplateMap = {};

  templates.forEach(({ files, script: templateScript, templateName }) => {
    const parsedFiles: ParsedTemplateFile[] = [];

    files.forEach((path) => {
      const content = fs.readFileSync(path, 'utf8');
      const parsed = parseAllInterpolationMarks(content, reInterpolation, config);
      parsedFiles.push({
        file: path,
        parsed,
        content,
      });
    });

    parsedTemplateMap[templateName] = {
      parsedFiles,
      templateScript,
    };
  });

  return parsedTemplateMap;
};
