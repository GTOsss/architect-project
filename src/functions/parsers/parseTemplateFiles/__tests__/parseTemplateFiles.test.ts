import { createResultsTable } from './testUtils';
import { getFunctionArgument, getFunctionName } from '../parseTemplateFragments.utils';
import { parseAllInterpolationMarks } from '../parseTemplateFiles';
import { interpolationRegExp, functionTemplate, miniTemplate } from './mock';
import { arcMockConfig } from '../../../../../__tests__/mock';

describe('parseTemplateFiles', () => {
  it('getFunctionName', () => {
    const result = getFunctionName('toCamelCase(name)');
    expect(result).toEqual('toCamelCase');
  });

  it('getFunctionArgument', () => {
    const result = getFunctionArgument('toCamelCase(name)');
    expect(result).toEqual('name');
  });

  it('isCallFunctionStr', () => {
    const result = createResultsTable();
    expect(result).toMatchSnapshot();
  });

  it('parseAllInterpolationMarks', () => {
    const result = parseAllInterpolationMarks(functionTemplate, interpolationRegExp, arcMockConfig);
    expect(result).toMatchSnapshot();
  });

  it('parseAllInterpolationMarks mini template', () => {
    const result = parseAllInterpolationMarks(miniTemplate, interpolationRegExp, arcMockConfig);
    expect({ input: miniTemplate, result }).toMatchSnapshot();
  });
});
