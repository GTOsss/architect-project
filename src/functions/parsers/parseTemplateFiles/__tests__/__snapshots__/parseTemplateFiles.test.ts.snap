// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseTemplateFiles isCallFunctionStr 1`] = `
"func(a);             → true
func(a1);            → true
func1(a)             → true
func1(a2)            → true
camelNameFn()        → true
CamelNameFn()        → true
camelNameFn(a)       → true
CamelNameFn(a)       → true
toCamelCase(name)    → true
func (a);            → true
func (a1);           → true
func1 (a)            → true
func1 (a2)           → true
func(3)              → true
a()                  → true
a1()                 → true

(a)          → false
(a1)         → false
fn(a b)      → false
func(a b c)  → false
fn(1a a)     → false
test         → false
test1        → false"
`;

exports[`parseTemplateFiles parseAllInterpolationMarks 1`] = `
[
  {
    "data": {
      "functionArgs": [
        "name",
      ],
      "functionName": "toCamelCase",
      "type": "callFunction",
    },
    "index": 14,
    "originStr": "{{toCamelCase(name)}}",
    "str": "toCamelCase(name)",
  },
  {
    "data": {
      "type": "variable",
      "variableName": "name",
    },
    "index": 60,
    "originStr": "{{name}}",
    "str": "name",
  },
  {
    "data": {
      "functionArgs": [
        "name",
      ],
      "functionName": "toLowerCamelCase",
      "type": "callFunction",
    },
    "index": 83,
    "originStr": "{{toLowerCamelCase(name)}}",
    "str": "toLowerCamelCase(name)",
  },
  {
    "data": {
      "functionArgs": [
        "name",
      ],
      "functionName": "toCamelCase",
      "type": "callFunction",
    },
    "index": 117,
    "originStr": "{{toCamelCase(name)}}",
    "str": "toCamelCase(name)",
  },
]
`;

exports[`parseTemplateFiles parseAllInterpolationMarks mini template 1`] = `
{
  "input": "so{{me}} "{{fn(name)}}"",
  "result": [
    {
      "data": {
        "type": "variable",
        "variableName": "me",
      },
      "index": 4,
      "originStr": "{{me}}",
      "str": "me",
    },
    {
      "data": {
        "functionArgs": [
          "name",
        ],
        "functionName": "fn",
        "type": "callFunction",
      },
      "index": 12,
      "originStr": "{{fn(name)}}",
      "str": "fn(name)",
    },
  ],
}
`;
